<a href="https://colab.research.google.com/github/ricedavis/CSCE580-Fall2025/blob/main/Quiz1/Q3_Code.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# a. Data Issues

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [7]:
import pandas as pd

In [18]:
df = pd.read_csv('/content/drive/MyDrive/Redacted-Department _ CAD Reconciliation.csv')

In [29]:
df.head()

XREF ID         DISPATCH UNIT  ... CALL COMPLETE SHIFT
0  2025107105  BAT111, E171, LDR175  ...  9/5/25 16:20     C
1  2025107223                BAT111  ...  9/5/25 17:55     C
2  2025107415                  E171  ...  9/4/25 21:09     C
3  2025107411                  E171  ...  9/4/25 21:02     C
4  2025107384  BAT111, E171, LDR175  ...  9/4/25 21:36     C

[5 rows x 8 columns]

## 1. Range of Data

In [34]:
df['DISPATCH CREATED DATE'] = pd.to_datetime(df['DISPATCH CREATED DATE'], errors = 'coerce')

In [33]:
# Calculate the range of data by taking the last entry time minus the first entry time
print(f"min: {df['DISPATCH CREATED DATE'].min()}, max: {df['DISPATCH CREATED DATE'].max()}")
time_range = df['DISPATCH CREATED DATE'].max() - df['DISPATCH CREATED DATE'].min()
time_range

min: 2025-03-24 15:54:00, max: 2025-08-31 23:03:00


Timedelta('160 days 07:09:00')

The range of data is from 2025-03-24 15:54:00 to 2025-08-31 23:03:00, the total change in time being 160 days, 7 hours and 9 minutes.

## 2. Percent of data missing from each column

In [39]:
# Find number of null values for each column
df.isna().sum()

XREF ID                    0
DISPATCH UNIT              0
DISPATCH CREATED DATE      0
INCIDENT NUMBER            0
1ST UNIT ON SCENE        428
ALARM DATE TIME           31
CALL COMPLETE             31
SHIFT                     69
dtype: int64

In [60]:
# Amount of values that should be in each column
len(df)

# Calculate percent of missing values per column
def calc_percent(column):
  percent_missing = df[column].isna().sum()/2200*100
  return f"{column} percent of missing values = {percent_missing:.2f}%"

for column in df:
  result = calc_percent(column)
  print(result)

XREF ID percent of missing values = 0.00%
DISPATCH UNIT percent of missing values = 0.00%
DISPATCH CREATED DATE percent of missing values = 0.00%
INCIDENT NUMBER percent of missing values = 0.00%
1ST UNIT ON SCENE percent of missing values = 19.45%
ALARM DATE TIME percent of missing values = 1.41%
CALL COMPLETE percent of missing values = 1.41%
SHIFT percent of missing values = 3.14%


Calculated the percent of missing values for each column, with the column '1ST UNIT ON SCENE' having the highest rate of missing values.

## 3. Identifying Issues in the Data

Based on the prior task, we know that there are lots of **missing values**, that is one issue with the data.

In [61]:
df.dtypes

XREF ID                           int64
DISPATCH UNIT                    object
DISPATCH CREATED DATE    datetime64[ns]
INCIDENT NUMBER                  object
1ST UNIT ON SCENE                object
ALARM DATE TIME                  object
CALL COMPLETE                    object
SHIFT                            object
dtype: object

In [63]:
df.head()

XREF ID         DISPATCH UNIT  ... CALL COMPLETE SHIFT
0  2025107105  BAT111, E171, LDR175  ...  9/5/25 16:20     C
1  2025107223                BAT111  ...  9/5/25 17:55     C
2  2025107415                  E171  ...  9/4/25 21:09     C
3  2025107411                  E171  ...  9/4/25 21:02     C
4  2025107384  BAT111, E171, LDR175  ...  9/4/25 21:36     C

[5 rows x 8 columns]

Some issues highlighted here are that some columns which represent dates are not in the proper datetime format for manipulation. "ALARM DATE TIME" and "CALL COMPLETE" are examples of these columns.

Other formatting issues can be found within the "DISPATCH UNIT" column, where some rows include different units seperated by commas. This inconsistency can make it tricky to analyze.

Logical errors may exist as well. It is important to check that the value in the "CALL COMPLETE" column does not exceed the "ALARM DATE TIME" as that wouldn't make sense in this context.

Finally, it's important to check for duplicates within the data as well.

## 4. Resolving Data Issues